examples/src/main/python/sql/streaming/structured_network_wordcount_session_window.py

```diff
@@ -0,0 +1,130 @@
+    examples/src/main/python/sql/streaming/structured_network_wordcount_session_window.py
+    localhost 9999`
+"""
+import sys
+import math
+
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import explode
+from pyspark.sql.functions import split
+from pyspark.sql.types import (
+    LongType,
+    StringType,
+    StructType,
+    StructField,
+)
+from pyspark.sql.streaming.state import GroupStateTimeout, GroupState
+import pandas as pd
+from typing import Iterable, Any
+
+if __name__ == "__main__":
+    if len(sys.argv) != 3:
+        msg = "Usage: structured_network_wordcount_session_window.py <hostname> <port>"
+        print(msg, file=sys.stderr)
+        sys.exit(-1)
+
+    host = sys.argv[1]
+    port = int(sys.argv[2])
+
+    spark = SparkSession.builder.appName(
+        "StructuredNetworkWordCountSessionWindow"
+    ).getOrCreate()
+
+    spark.sparkContext.setLogLevel("WARN")
+
+    # Create DataFrame representing the stream of input lines from connection to host:port
+    lines = (
+        spark.readStream.format("socket")
+        .option("host", host)
+        .option("port", port)
+        .option("includeTimestamp", "true")
+        .load()
+    )
+
+    # Split the lines into words, retaining timestamps
+    # split() splits each line into an array, and explode() turns the array into multiple rows
+    events = lines.select(
+        explode(split(lines.value, " ")).alias("sessionId"),
+        lines.timestamp.cast("long"),
+    )
+    session_fields = [
+        StructField("sessionId", StringType()),
+        StructField("count", LongType()),
+        StructField("start", LongType()),
+        StructField("end", LongType()),
+    ]
+
+    def func(
+        key: Any, pdf_iter: Iterable[pd.DataFrame], state: GroupState
+    ) -> Iterable[pd.DataFrame]:
+        if state.hasTimedOut:
+            count, start, end = state.get
+            state.remove()
+            yield pd.DataFrame(
+                {
+                    "sessionId": [key[0]],
+                    "count": [count],
+                    "start": [start],
+                    "end": [end],
+                }
+            )
+        else:
+            start = math.inf
+            end = 0
+            count = 0
+            for pdf in pdf_iter:
+                start = min(start, min(pdf["timestamp"]))
+                end = max(end, max(pdf["timestamp"]))
+                count = count + len(pdf)
+            if state.exists:
+                old_session = state.get
+                count = count + old_session[0]
+                start = old_session[1]
+                end = max(end, old_session[2])
+            state.update((count, start, end))
+            state.setTimeoutDuration(10000)
+            yield pd.DataFrame()
+
+    # Group the data by window and word and compute the count of each group
+    sessions = events.groupBy(events["sessionId"]).applyInPandasWithState(
+        func,
+        StructType(session_fields),
```

Comment:
Can we have two separate variables for these schemas?

Reply:
I prefer to keep it because all fields are the same except the first, make the code less verbose.

Reply:
Since this is an example (for newbies), let's make it verbose.

Reply:
Makes sense, I just make the code more readable, PTAL.